cfg.c

```diff
@@ -0,0 +1,245 @@
+}
+
+static int slurp(int fd, size_t to_read, char **dst) {
+  char *buffer, *w;
+
+  buffer = malloc(to_read + 1);
+  if (!buffer)
+    return PAM_BUF_ERR;
+
+  w = buffer;
+  while (to_read) {
+    ssize_t r;
+
+    r = read(fd, w, to_read);
+    if (r < 0) {
+      free(buffer);
+      return PAM_SYSTEM_ERR;
+    }
+
+    if (r == 0)
+      break;
+
+    w += r;
+    to_read -= r;
+  }
+
+  *w = '\0';
+  *dst = buffer;
+  return PAM_SUCCESS;
+}
+
+static const char *ltrim(const char *s) {
+  while (isspace((unsigned char) *s))
+    s++;
+  return s;
+}
+
+static int cfg_load_defaults(cfg_t *cfg, const char *config_path) {
+  int fd, e;
+  struct stat st;
+  char *buffer = NULL, *saveptr = NULL;
+  const char *arg;
+
+  // If a config file other than the default is provided,
+  // the path must be absolute.
+  if (config_path && *config_path != '/')
+    return PAM_AUTHINFO_UNAVAIL;
+
+  fd = open(config_path ? config_path : CFG_DEFAULT_PATH,
+            O_RDONLY | O_CLOEXEC | O_NOCTTY | O_NOFOLLOW, 0);
+  if (fd == -1) {
+
+    // Only the default config file is allowed be missing.
+    if (errno == ENOENT && !config_path)
+      return PAM_SUCCESS;
+
+    return PAM_AUTHINFO_UNAVAIL;
+  }
+
+  if (fstat(fd, &st)) {
+    e = PAM_SYSTEM_ERR;
+    goto exit;
+  }
+
+  if (!S_ISREG(st.st_mode)) {
+    e = PAM_AUTHINFO_UNAVAIL;
+    goto exit;
+  }
+
+  if (st.st_size == 0) {
+    e = PAM_SUCCESS;
+    goto exit;
+  }
+
+  if (st.st_size > CFG_MAX_FILE_SIZE) {
+    e = PAM_AUTHINFO_UNAVAIL;
+    goto exit;
+  }
+
+  e = slurp(fd, st.st_size, &buffer);
+  if (e)
+    goto exit;
+
+  arg = strtok_r(buffer, "\n", &saveptr);
+  while (arg) {
+    arg = ltrim(arg);
+
+    if (arg[0] != '\0' && arg[0] != '#') {
+      e = cfg_load_arg(cfg, arg);
+      if (e)
+        goto exit;
+    }
+
+    arg = strtok_r(NULL, "\n", &saveptr);
+  }
+
+  // Transfer buffer ownership
+  cfg->defaults_buffer = buffer;
+  buffer = NULL;
+  e = PAM_SUCCESS;
+
+exit:
+  free(buffer);
+  close(fd);
+  return e;
+}
+
+static void cfg_reset(cfg_t *cfg) {
+  memset(cfg, 0, sizeof(cfg_t));
+  cfg->debug_file = DEFAULT_DEBUG_FILE;
+  cfg->userpresence = -1;
+  cfg->userverification = -1;
+  cfg->pinverification = -1;
+}
+
+int cfg_init(cfg_t *cfg, int flags, int argc, const char **argv) {
+  int i, e;
+  const char *config_path = NULL;
+
+  cfg_reset(cfg);
+
+  // Config file needs to be pre-loaded.
+  // If multiple conf_file= options appear, the last is honoured
+  // (consistently with any other option).
+  for (i = argc - 1; i >= 0; i--) {
+    if (strncmp(argv[i], "conf_file=", strlen("conf_file=")) == 0) {
+      config_path = argv[i] + strlen("conf_file=");
+      break;
+    }
+  }
+
+  e = cfg_load_defaults(cfg, config_path);
+  if (e != PAM_SUCCESS)
+    goto fail;
+
+  for (i = 0; i < argc; i++) {
+    if (strncmp(argv[i], "conf_file=", strlen("conf_file=")) == 0)
+      continue;
+    e = cfg_load_arg(cfg, argv[i]);
+    if (e != PAM_SUCCESS)
+      goto fail;
```

Comment:
Not sure if we want to reject here.
Doing so is probably safer in case the of mistake by the sysadmin, but it would also introduce some breaking behavior. Any thought?